Add live list of pods to PVC protection controller to make sure it does not delete a PVC which is being used by a Pod

matte21 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

The PVC protection controller processes PVCs with `deletionTimestamp` set to assess whether its finalizer can be removed (to enable deletion of the PVC). To do that, it checks whether there exists a Pod using the PVC (can't be deleted) or not (can be deleted). Prior to this PR, the controller retrieved all the pods with an Informer's cache-based list. This is incorrect because a Pod using the PVC might not be in the cache yet. If that's the case a PVC can be deleted while being used by a Pod. This PR fixes this by retrieving all the pods with a live list (which is guaranteed to return all existing pods) after the cache-based one. 

Note that it would be enough to do only the live list. Instead, this PR adds it after the cache-based list and performs it only if the latter finds no Pod blocking deletion of the PVC being processed. The
rationale is that live lists are expensive and it's desirable to minimize them. The drawback is that if at the time of the cache-based list the cache has not been notified yet of the deletion of a Pod using
the PVC the PVC is kept. Correctness is not compromised because the finalizer will be removed when the Pod deletion notification is received, but this means PVC deletion is delayed. Reducing live lists was valued more than deleting PVCs slightly faster.

Also, currently the two lists retrieve all the pods in the same namespace as the PVC being processed. Further optimizations based on adding field selectors to reduce the number of fetched pods are possible but left for the future (in this PR or others).

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #75980

**Special notes for your reviewer**:


**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [Original PVC protection controller design doc]: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md 
```

matte21 committed:
Add live list of pods to PVC protection controller

Add live list of pods to PVC protection controller, as opposed to doing
only a cache-based list through the Informer. Both lists are performed
while processing a PVC with deletionTimestamp set to check whether Pods
using the PVC exist and remove the finalizer to enable deletion of the
PVC if that's not the case. Prior to this commit only the cache-based
list was done but that's unreliable because a pod using the PVC might
exist but not be in the cache just yet. On the other hand, the live
list is 100% reliable.

Note that it would be enough to do only the live list. Instead, this
commit adds it after the cache-based list and performs it only if the
latter finds no Pod blocking deletion of the PVC being processed. The
rationale is that live lists are expensive and it's desirable to
minimize them. The drawback is that if at the time of the cache-based
list the cache has not been notified yet of the deletion of a Pod using
the PVC the PVC is kept. Correctness is not compromised because the
finalizer will be removed when the Pod deletion notification is
received, but this means PVC deletion is delayed. Reducing live lists
was valued more than deleting PVCs slightly faster.

Also, add a unit test that fails without the change introduced by this
commit and revamp old unit tests. The latter is needed because expected
behavior is described in terms of API calls the controller makes, and
this commit introduces new API calls (the live lists).

matte21 commented:
/sig storage

msau42 commented:
/ok-to-test
@kubernetes/sig-storage-pr-reviews

cwdsuzhou commented on pkg/controller/volume/pvcprotection/pvc_protection_controller.go:
```diff
@@ -209,42 +211,86 @@ func (c *Controller) removeFinalizer(pvc *v1.PersistentVolumeClaim) error {
-				klog.V(2).Infof("Keeping PVC %s/%s, it is used by pod %s/%s", pvc.Namespace, pvc.Name, pod.Namespace, pod.Name)
-				return true, nil
+			if volume.PersistentVolumeClaim != nil && volume.PersistentVolumeClaim.ClaimName == pvc {
+				klog.V(2).Infof("Pod %s/%s uses PVC %s", pod.Namespace, pod.Name, pvc)
```
IMO keep old code style may be better 
```
if volume.PersistentVolumeClaim == nil {
    continue
}
if volume.PersistentVolumeClaim.ClaimName == pvc.Name{
    ...
}
```

matte21 replied:
I'd rather leave it unchanged, unless your opinion has consensus among the reviewers or you can provide a Code style doc that suggests the old style is preferred.

cwdsuzhou commented on pkg/controller/volume/pvcprotection/pvc_protection_controller.go:
```diff
@@ -209,42 +211,86 @@ func (c *Controller) removeFinalizer(pvc *v1.PersistentVolumeClaim) error {
+	// mean such a Pod doesn't exist: it might just not be in the cache yet. To
+	// be 100% confident that it is safe to delete pvc make sure no Pod is using
+	// it among those returned by a live list.
+	return c.askAPIServer(pvc)
```
We may need evaluate the load increase since this func may be frequently called when when we delete pvcs.

MikeSpreitzer replied:
There is a correct way to reduce that load.  See the "A correct optimization" section in https://www.ibm.com/cloud/blog/new-builders/preserve-kubernetes-api-objects-while-they-are-in-use?mhsrc=ibmsearch_a&mhq=preserve%20kubernetes%20api%20object

matte21 replied:
That optimization is already part of this PR.

MikeSpreitzer replied:
Sorry, did not read enough.

matte21 commented:
I rebased on master to get in the changes on the same files introduced by #80476 and avoid future merge conflicts. I will be away this weekend, work on this will resume on Monday.

matte21 commented:
/needs-ok-to-test

matte21 commented:
/assign @jsafrane

matte21 commented:
So as I wrote above I have doubts on how to modify unit tests (simply "tests" from now on) in the light of the changes introduced in this PR, which is why it's marked as WIP. Note that this PR fixes a bug, hence at least a new test case which fails without the introduced change must be added. My doubt is not on how to do that. It's about how to keep the unit tests decoupled from the implementation --- the introduced change makes it harder. Follows the description of the problem and a proposed solution. The latter is not part of this PR yet (it might even end up in its own PR if deemed appropriate), I'd like it to reach some consensus before coding it (if ever).

Currently tests outcome is determined by comparing the expected actions to those created with the fake client (which records and exposes them). It's a success if the two sequences are identical, failure otherwise. Action means API Server Call. Prior to this PR the only actions the controller generated were explicit in its behavior: updates to PVC status to add/remove the finalizer. After this PR there are live lists as well, which are an implementation detail, not part of the controller's behavior and IMO should not be part of the expected tests result because this couples the tests with the implementation. But the way current tests are designed pushes toward this coupling.

IMO tests should be refactored to avoid the aforementioned coupling. My proposal is to assess tests outcome by comparing the content of the object tracker acting as the underlying store of the fake client to the expected set of API objects. Follows an example. Consider the test case where the only existing API object is a PVC whose `metadata.deletionTimestamp` gets set. The controller has to remove the finalizer. Using the current tests the expected result would consist of a slice whose elements are a live list of Pods and an update action (which wraps the PVC object without the finalizer). The test outcome would be determined by making sure such slice is equal to that of the actions executed on the fake client. After the proposed change, the test expected result would simply be the PVC object without the finalizer, and the test outcome would be decided by checking whether after some time the object tracker acting as the underlying store of the fake client contains only the PVC without the finalizer and nothing else.

matte21 commented:
Can someone remove the `/ok-to-test`label? Currently most tests are meant to fail.

msau42 commented:
/ok-to-test

matte21 commented:
/help

jsafrane commented:
> After this PR there are live lists as well, which are an implementation detail, not part of the controller's behavior and IMO should not be part of the expected tests result because this couples the tests with the implementation.

These tests are unit tests and they're white-box by definition. They test the implementation. I would not be afraid to hardcode any API calls / responses into the tests.

> My proposal is to assess tests outcome by comparing the content of the object tracker acting as the underlying store of the fake client to the expected set of API objects. Follows an example. Consider the test case where the only existing API object is a PVC whose metadata.deletionTimestamp gets set. The controller has to remove the finalizer. Using the current tests the expected result would consist of a slice whose elements are a live list of Pods and an update action (which wraps the PVC object without the finalizer). The test outcome would be determined by making sure such slice is equal to that of the actions executed on the fake client. 

I am not sure I understand it completelly, but check PV controller test and [its framework](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/framework_test.go#L660), it simulates API calls until all objects reach expected final state. But it's fairly complicated and IMO not necessary in this case.

matte21 commented:
> it simulates API calls until all objects reach expected final state

That's what I meant.

> These tests are unit tests and they're white-box by definition. They test the implementation. I would not be afraid to hardcode any API calls / responses into the tests.

Ok, then I'll take this approach since it's not a problem.

matte21 commented:
df79143 includes both the bug fix and the addition and revamp of the unit tests. This PR is now complete.

jsafrane commented on pkg/controller/volume/pvcprotection/pvc_protection_controller.go:
```diff
@@ -209,42 +211,86 @@ func (c *Controller) removeFinalizer(pvc *v1.PersistentVolumeClaim) error {
 	pods, err := c.podLister.Pods(pvc.Namespace).List(labels.Everything())
 	if err != nil {
-		return false, err
+		return false, fmt.Errorf("Cache-based list of pods failed while processing %s/%s: %s", pvc.Namespace, pvc.Name, err.Error())
```
Errors should start with lowercase so they can be chained easily (just like you do here). 
This applies to all `fmt.Errorf` in this PR.

jsafrane replied:
Hmm, I can see the original file has it wrong everywhere. Do you mind fixing it (as a separate commit)?

matte21 replied:
Done in bf0f3d3.

In case you plan to merge this PR with the two commits unsquashed I can reword the second commit's message to make it clear that the change affects the PVC protection controller.

jsafrane reviewed (commented):
It looks great, good work.

matte21 committed:
Lowercase first letter of chainable errors in PVC protection controller

Lowercase first letter of error messages that are not printed right
away to ease chaining with other error messages in PVC protection
controller.

matte21 commented:
/test pull-kubernetes-integration

matte21 commented:
/test pull-kubernetes-bazel-build

matte21 commented:
/test pull-kubernetes-e2e-gce-csi-serial

jsafrane commented on pkg/controller/volume/pvcprotection/pvc_protection_controller.go:
```diff
@@ -209,44 +211,88 @@ func (c *Controller) removeFinalizer(pvc *v1.PersistentVolumeClaim) error {
+}
+
+func podUsesPVC(pod *v1.Pod, pvc string) bool {
+	klog.V(5).Infof("Checking whether PVC %s/%s is used by Pod %s", pod.Namespace, pvc, pod.Name)
```
Why `V(5)`? Rest of the code uses `V(4)`

matte21 replied:
The rest of the code that uses `V(4)` logs stuff that happens at most once while processing a PVC, while that log statement can be executed multiple times when processing a PVC and IMO it is about a lower-level detail.

jsafrane commented on pkg/controller/volume/pvcprotection/pvc_protection_controller.go:
```diff
@@ -209,44 +211,88 @@ func (c *Controller) removeFinalizer(pvc *v1.PersistentVolumeClaim) error {
 
-	klog.V(3).Infof("PVC %s/%s is unused", pvc.Namespace, pvc.Name)
-	return false, nil
+	klog.V(5).Infof("PVC %s/%s is not used by Pod %s", pod.Namespace, pvc, pod.Name)
```
I don't think we need to log it per pod, that's just too much of logs.

matte21 replied:
That's why the verbosity is higher than that of the other log statements. I'm ok both with removing it or keeping it (and optionally increasing its verbosity to 6), but what about the other `V(5)` log statement (the one your previous comment is about)? Should I remove that as well or keep only that one?

jsafrane replied:
I'd prefer all per-pod logs to be removed, IMO the overall result (used by pod XYZ / unused) should be enough.

jsafrane reviewed (commented):
Good work with the unit test. I just have few logging nits.

jsafrane commented:
/lgtm
/approve

Thanks for this PR!

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest

matte21 commented:
@jsafrane thanks for your guidance! Should I squash all commit?

matte21 commented:
/hold

jsafrane commented:
Yes, please squash the first and the last commit.
/lgtm cancel
/hold cancel
(will this work?)

matte21 commented:
/retest

matte21 commented:
> Yes, please squash the first and the last commit.

@jsafrane done.

jsafrane commented:
/lgtm
